File: 03. Stack and Queue/01. Valid parentheses/main.cpp
#include <bits/stdc++.h>
using namespace std;

bool isValid(string s) {
    stack<char> st;

    for (char c : s) {
        if (c == '(' || c == '{' || c == '[') {
            st.push(c); // Push opening brackets onto stack
        } else {
            if (st.empty()) return false; // No matching opening bracket

            char top = st.top();
            if ((c == ')' && top == '(') || 
                (c == '}' && top == '{') || 
                (c == ']' && top == '[')) {
                st.pop(); // Valid pair, remove from stack
            } else {
                return false; // Mismatched closing bracket
            }
        }
    }

    return st.empty(); // Stack should be empty if all brackets are matched
}

int main() {
    vector<string> testCases = {"()", "()[]{}", "(]", "([)]", "{[]}"};

    for (const string &test : testCases) {
        cout << "String: " << test << " -> " << (isValid(test) ? "Valid" : "Invalid") << endl;
    }

    return 0;
}

File: 03. Stack and Queue/01. Valid parentheses/main.js
function isValid(s) {
    let stack = [];

    for (let c of s) {
        if (c === '(' || c === '{' || c === '[') {
            stack.push(c); // Push opening brackets
        } else {
            if (stack.length === 0) return false; // No matching opening bracket

            let top = stack[stack.length - 1];
            if ((c === ')' && top === '(') || 
                (c === '}' && top === '{') || 
                (c === ']' && top === '[')) {
                stack.pop(); // Valid pair, remove from stack
            } else {
                return false; // Mismatch
            }
        }
    }
    return stack.length === 0; // Stack should be empty if all brackets match
}


function main() {
    const testCases = ["()", "()[]{}", "(]", "([)]", "{[]}"];
    testCases.forEach(test => {
        console.log(`String: ${test} -> ${isValid(test) ? "Valid" : "Invalid"}`);
    });
}

main();